File: code/Java/src/main/java/Infrastructure/SpaceComm/MarsRoverReceiver.java
package Infrastructure.SpaceComm;

import Infrastructure.Bus.IMessageReceivedBus;
import Infrastructure.INotifier;
import Infrastructure.Bus.IWriteToServiceBus;
import Infrastructure.Timing.IWaitMillisecons;
import Infrastructure.Timing.SmartTimer;

import java.util.ArrayList;

public class MarsRoverReceiver implements IWriteToServiceBus {

    private IMessageReceivedBus marsRoverServiceBus;
    private ArrayList<String> datagrams = new ArrayList<>();
    private IWaitMillisecons smartTimer = SmartTimer.New();

    private INotifier notifyMessage = (data) -> this.notifyMessage(data);
    private int maxDelay;

    public MarsRoverReceiver(int maxDelay) {

        this.maxDelay = maxDelay;
    }

    public void writesTo(IMessageReceivedBus marsRoverServiceBus) {

        this.marsRoverServiceBus = marsRoverServiceBus;
    }

    public void received(String datagram) {
        datagrams.add(datagram);

        Message message = new Message(datagrams);

        if (message.isValid()) {
            marsRoverServiceBus.NotifyMessageReceived(message.toString());
            return;
        }

        smartTimer
                .waitMillisecond(maxDelay)
                .beforeDoing(notifyMessage, this.datagrams);
    }

    private void notifyMessage(ArrayList<String> datagrams) {
        Message message = new Message(datagrams);
        if (message.isValid()) {
            marsRoverServiceBus.NotifyMessageReceived(message.toString());
            return;
        }
        marsRoverServiceBus.NotifyError();
    }

}

File: code/CSharp/ConnascentMarsRover/Source/Infrastructure/SpaceComm/MarsRoverReceiver.cs
using System.Collections.Generic;
using Source.Infrastructure.Bus;
using Source.Infrastructure.Timing;

namespace Source.Infrastructure.SpaceComm
{
    public class MarsRoverReceiver : IWriteToServiceBus
    {
        private const int MaxDelayMilliseconds = 3000;
        private IMessageReceivedBus _marsRoverServiceBus;
        private readonly List<string> _packets = new List<string>();
        private readonly ISmartTimer _smartTimer = new SmartTimer();

        public void Received(string packet)
        {
            _packets.Add(packet);
            var message = new Message(_packets);
            if (message.IsValid())
            {
                _marsRoverServiceBus.NotifyMessageReceived(message.ToString());
                return;
            }

            _smartTimer
                .WaitMilliseconds(MaxDelayMilliseconds)
                .BeforeDoing(NotifyMessage, _packets);
        }

        public void WritesTo(IMessageReceivedBus messageReceivedBus)
        {
            _marsRoverServiceBus = messageReceivedBus;
        }

        private void NotifyMessage(List<string> packets)
        {
            var message = new Message(packets);
            if (message.IsValid())
            {
                _marsRoverServiceBus.NotifyMessageReceived(message.ToString());
                return;
            }

            _marsRoverServiceBus.NotifyError();
        }
    }
}